src/repository_urls.rs

```diff
@@ -55,261 +48,150 @@ fn get_distro_name(sysinfo: &SystemInfo, distro: &str) -> Option<String> {
+///
+/// Thus the full path segment is `__linux__/<distribution codename>/<snapshot date>/src/contrib/<file name>?r_version=<R version major>.<R version minor>&arch=<system arch>`
+fn get_binary_path(url: &str, file_name: &str, r_version: &[u32; 2], sysinfo: &SystemInfo) -> Option<String> {
+    // rv does not support binaries for less than R/3.6
+    if r_version < &[3, 6] {
+        return None;
     }
 
-    fn get_binary_tarball_path(
-        &self,
-        name: &str,
-        version: &str,
-        path: Option<&str>,
-        r_version: &[u32; 2],
-        sysinfo: &SystemInfo,
-    ) -> Option<String> {
-        let ext = sysinfo.os_type.tarball_extension();
-        let p = if let Some(p2) = path {
-            format!("{p2}/")
-        } else {
-            String::new()
-        };
-        let file_name = format!("{p}{name}_{version}.{ext}");
-        self.get_binary_path(&file_name, r_version, sysinfo)
-    }
-
-    fn get_source_path(&self, file_name: &str) -> String {
-        let url = self.url();
-        format!("{url}/src/contrib/{file_name}")
-    }
-
-    fn get_source_tarball_path(&self, name: &str, version: &str, path: Option<&str>) -> String {
-        let p = if let Some(p2) = path {
-            format!("{p2}/")
-        } else {
-            String::new()
-        };
-        let file_name = format!("{p}{name}_{version}.tar.gz");
-        self.get_source_path(&file_name)
-    }
-
-    // Archived packages under the format <base url>/src/contrib/Archive/<pkg name>/<pkg name>_<pkg version>.tar.gz
-    fn get_archive_tarball_path(&self, name: &str, version: &str) -> String {
-        let file_name = format!("Archive/{name}/{name}_{version}.tar.gz");
-        self.get_source_path(&file_name)
+    match sysinfo.os_type {
+        OsType::Windows => Some(get_windows_url(url, file_name, r_version)),
+        OsType::MacOs => get_mac_url(url, file_name, r_version, sysinfo),
+        OsType::Linux(distro) => get_linux_url(url, file_name, r_version, sysinfo, distro),
+        OsType::Other(_) => None,
     }
+}
 
-    fn get_tarball_urls(
-        &self,
-        name: &str,
-        version: &str,
-        path: Option<&str>,
-        r_version: &[u32; 2],
-        sysinfo: &SystemInfo,
-    ) -> TarballUrls {
-        let source = self.get_source_tarball_path(name, version, path);
-        let binary = self.get_binary_tarball_path(name, version, path, r_version, sysinfo);
-        let archive = self.get_archive_tarball_path(name, version);
-        TarballUrls {
-            source,
-            binary,
-            archive,
-        }
-    }
+fn get_binary_tarball_path(
+    url: &str, 
+    name: &str,
+    version: &str,
+    path: Option<&str>,
+    r_version: &[u32; 2],
+    sysinfo: &SystemInfo,
+) -> Option<String> {
+    let file_name = format!(
+        "{}{}_{}.{}",
+        path.map(|x| format!("{x}/")).unwrap_or_default(),
+        name,
+        version,
```

Comment:
pure cosmetic but:

```
let path = path.map(|x| format!("{x}/")).unwrap_or_default();
let extension = sysinfo.os_type.tarball_extension();

let file_name = format!("{path}{name}_{version}.{extension}");
```
feels easier to understand. Also if the path can have `/` it's not really a filename I'd say